workers/proto/kitchen_sink/kitchen_sink.proto

```diff
@@ -0,0 +1,328 @@
+}
+
+message DoQuery {
+  oneof variant {
+    // A query must exist named `report_state` which returns the `WorkflowState` message. The input
+    // is pointless and only exists to allow testing of variably-sized query args.
+    temporal.api.common.v1.Payloads report_state = 1;
+    // Send an arbitrary query
+    HandlerInvocation custom = 2;
+  }
+}
+
+message DoUpdate {
+  oneof variant {
+    // An update must exist named `do_actions_update` which does the same thing the signal handler
+    // would do when given the provided action set. The handler should return the `WorkflowState`
+    // when done.
+    ActionSet do_actions = 1;
+    // This update request must be rejected by the workflow using a handler named `always_reject`
+    google.protobuf.Empty reject_me = 2;
+    // This update request must be accepted and then failed by the workflow using a handler named
+    // `always_fail`
+    temporal.api.failure.v1.Failure fail_update = 3;
+    // Send an arbitrary update request
+    HandlerInvocation custom = 4;
+  }
+}
+
+message HandlerInvocation {
+  string name = 1;
+  repeated temporal.api.common.v1.Payload args = 2;
+}
+
+// Each workflow must maintain an instance of this state
+message WorkflowState {
+  map<string, string> kvs = 1;
+}
+
+message WorkflowInput {
+  repeated ActionSet initial_actions = 1;
+}
+
+// A set of actions to execute concurrently or sequentially. It is necessary to be able to represent
+// sequential execution without multiple 1-size action sets, as that implies the receipt of a signal
+// between each of those sets, which may not be desired.
+//
+// All actions are handled before proceeding to the next action set, unless one of those actions
+// would cause the workflow to complete/fail/CAN.
+message ActionSet {
+  repeated Action actions = 1;
+  bool concurrent = 2;
+}
+
+message Action {
+  oneof variant {
+    TimerAction timer = 1;
+    ExecuteActivityAction exec_activity = 2;
+    ExecuteChildWorkflowAction exec_child_workflow = 3;
+    AwaitSignalAction await_signal = 4;
+    AwaitUpdateAction await_update = 5;
+    SendSignalAction send_signal = 6;
+    SetPatchMarkerAction set_patch_marker = 7;
+    UpsertSearchAttributesAction upsert_search_attributes = 8;
+    UpsertMemoAction upsert_memo = 9;
+    WorkflowState set_workflow_state = 10;
+
+    ReturnResultAction return_result = 11;
+    ReturnErrorAction return_error = 12;
+    ContinueAsNewAction continue_as_new = 13;
+
+    ActionSet nested_action_set = 14;
+  }
+}
+
+// The following actions and the messages they depend on are largely duplicative of Core/Server's
+// commands, but some care is taken to remove things that would be uninteresting for these tests.
+
+// All cancellable commands will have this available as a field. If it is set, the command
+// should be cancelled at the specified juncture.
+message CancelCondition {
+  oneof condition {
+    google.protobuf.Empty before_started = 1;
+    google.protobuf.Empty after_started = 2;
```

Comment:
How does  this work with activities since at least Go and Java do not expose an activity has started  to users.

Reply:
Good question - I will add a comment. The idea is that the SDK will do this by distinguishing between cancelling _immediately_ (same WFT), or cancelling it in a subsequent WFT but before completion. A bit of thought might be needed here on how to make that work consistently from the generator / workflow impls.